cpp/include/data_structure/tree/fenwick_tree.hpp

```diff
@@ -0,0 +1,158 @@
+/*
+    Fenwick Tree
+    ------------
+    A fenwick tree is conceptually a tree, but it is implemented as a flat array
+    that allows the quick calculations prefix sums in an array. It accomplishes this by 
+    storing sums according to the binary representations of the indexes. For practical
+    purposes the fenwick tree is indexed starting at 1. 
+    Below is a more in-depth explanation of how the tree is constructed.
+    ...............................
+    start of sum         end of sum
+    (index - 2^r + 1) -> index 
+    ...............................
+    Where r is the last significant bit of the binary representation of index.
+    For example:
+    index = 6, 110 (in binary)
+    r = 1
+    (6 - 2^1 + 1) -> 6
+    5 -> 6
+    
+    So the sum of the value stored a 5 through 6 will be stored at index 6. Now 
+    remember the index is incremented by one from the original array. So at 
+    index = 6, of the fenwick tree, the sum of 4 through 5, of the original array
+    will be stored.
+    
+    This set up allows any prefix sum to be calculated in at most log(n) operations,
+    where n is the number of elements in the original array. To update the tree
+    it takes at most log(n) operations as well.
+*/
+
+#ifndef FENWICK_TREE_HPP
+#define FENWICK_TREE_HPP
+
+#include <vector>
+
+using std::vector;
+
+class FenwickTree {
+private:
+  vector<int> tree;
+  
+  int max_index;
+
+public:
+  FenwickTree(vector<int> original_array);
+  
+  void update_tree(int value, int index);
```

Comment:
It's odd to me that the value is passed before the index. I would've expected the opposite. That said, I'll defer to you since you clearly know way more about this than I do.